lib/utils/log/writer.go

```diff
@@ -43,3 +50,85 @@ func (s *SharedWriter) Write(p []byte) (int, error) {
 func NewSharedWriter(w io.Writer) *SharedWriter {
 	return &SharedWriter{Writer: w}
 }
+
+// FileSharedWriter is similar to SharedWriter except that it requires a os.File instead of a io.Writer.
+// This is to allow the File reopen required by logrotate and similar tools.
+// SharedWriter must be used for log destinations that don't have the reopen requirement, like stdout and stderr.
+// This is thread safe.
+type FileSharedWriter struct {
+	*os.File
+	fileFlag int
+	fileMode fs.FileMode
+	mu       sync.Mutex
+}
+
+func (s *FileSharedWriter) Write(p []byte) (int, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.File.Write(p)
+}
+
+// Reopen closes the file and opens it again using APPEND mode.
+func (s *FileSharedWriter) Reopen() (err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if err := s.Close(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.File, err = os.OpenFile(s.Name(), s.fileFlag, s.fileMode)
+	return trace.Wrap(err)
+}
+
+// runWatcher spawns goroutine with the watcher loop to consume events of moving
+// or removing the log file to re-open it for log rotation purposes.
+func (s *FileSharedWriter) runWatcher() error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	err = watcher.Add(s.Name())
```

Comment:
Are we guaranteed to get all relevant events by watching the file itself, or should we be watching the directory instead?

Reply:
By default `fsnotify` subscribes to all events (`Write`, `Remove`, `Rename`, `Chmod`), `Create` event should work in scope of directory only. 

Btw if we start watching the file and rename it, we stop receiving events. So might be watching directory be better option